test/boost/cdc_generation_test.cc

```diff
@@ -0,0 +1,118 @@
+/*
```

Comment:
This is a good opportunity to add a test that checks some invariants that generations should satisfy. We could consider only invariants that will not be broken by limiting the number of streams.
Some ideas:
1. each stream ID's token should be inside the token range designated by `prev_entry.end` and `entry.end`, where `entry` is the token range description containing this ID and `prev_entry` is the previous description in the vector (or the last if `entry` is the first one).
2. each token should be mapped to a stream ID in the same token range of the generation. That is, take the algorithm from cdc/metadata.cc and the generation (can be truncated), take a bunch of tokens (can be randomly generated), for each token perform this algorithm, and see if that is true. The algorithm should use the shard_counts given by the generation (and not e.g. smp::count), as it originally does. So, if the generation is truncated, the algorithm will use the truncated shard_count.
3. (optional/controversial) each shard should have at least one stream ID, even with truncated generations. That is, take each stream ID and calculate its shard (this time using the real smp::count!); see if each shard got at least one stream ID. This way could help us catch an obvious error such as placing all data on a strict subset of shards by wrongly truncating generations (the previous version of the PR did this).
Now this doesn't have to be true in theory I think (but I may be wrong) - perhaps there exists a choice of vnodes and a reduction of shard_count such that some shards won't get any stream ID (in any vnode). But I think it's highly improbable, if not impossible. However, if we're afraid of test flakiness, we could use a hardcoded seed to choose tokens deterministically (but run the test for a bunch of choices).